agent/lib_php_amqplib.c

```diff
@@ -0,0 +1,401 @@
+      .library = RABBITMQ_LIBRARY_NAME,
+      .destination_type = NR_MESSAGE_DESTINATION_TYPE_EXCHANGE,
+      .message_action = NR_SPANKIND_PRODUCER,
+      .messaging_system = RABBITMQ_MESSAGING_SYSTEM,
+  };
+
+  (void)wraprec;
+
+  amqp_exchange = nr_php_get_user_func_arg(2, NR_EXECUTE_ORIG_ARGS);
+  if (nr_php_is_zval_non_empty_string(amqp_exchange)) {
+    /*
+     * In PHP 7.x, the following will create a strdup in
+     * message_params.destination_name that needs to be freed.
+     */
+    message_params.destination_name
+        = ENSURE_PERSISTENCE(Z_STRVAL_P(amqp_exchange));
+  } else {
+    /* For producer, this is exchange name.  Exchange name is Default in case of
+     * empty string. */
+    message_params.destination_name = ENSURE_PERSISTENCE("Default");
+  }
+
+  amqp_routing_key = nr_php_get_user_func_arg(3, NR_EXECUTE_ORIG_ARGS);
+  if (nr_php_is_zval_non_empty_string(amqp_routing_key)) {
+    /*
+     * In PHP 7.x, the following will create a strdup in
+     * message_params.messaging_destination_routing_key that needs to be freed.
+     */
+    message_params.messaging_destination_routing_key
+        = ENSURE_PERSISTENCE(Z_STRVAL_P(amqp_routing_key));
+  }
+
+  amqp_connection = nr_php_get_zval_object_property(
+      nr_php_execute_scope(execute_data), "connection");
+  /*
+   * In PHP 7.x, the following will create a strdup in
+   * message_params.server_address that needs to be freed.
+   */
+  nr_php_amqplib_get_host_and_port(amqp_connection, &message_params);
+
+  /* For PHP 7.x compatibility. */
+  NR_PHP_WRAPPER_CALL
+
+  /*
+   * Now create and end the instrumented segment as a message segment.
+   *
+   * By this point, it's been determined that this call will be instrumented so
+   * only create the message segment now, grab the parent segment start time,
+   * add our message segment attributes/metrics then close the newly created
+   * message segment.
+   */
+
+  if (NULL == auto_segment) {
+    /*
+     * Must be checked after PHP_WRAPPER_CALL to ensure txn didn't end during
+     * the call.
+     */
+    goto end;
+  }
+  message_segment = nr_segment_start(NRPRG(txn), NULL, NULL);
+
+  if (NULL != message_segment) {
+    /* re-use start time from auto_segment started in func_begin */
+    message_segment->start_time = auto_segment->start_time;
+
+    nr_segment_message_end(&message_segment, &message_params);
+  }
+
+end:
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO
+    /* PHP 8+ */
+    /* gnu compiler needed closure. */
+    ;
+#else
+  /*
+   * Because we had to strdup values to persist them beyond NR_PHP_WRAPPER_CALL,
+   * now we destroy them. There isn't a separate function to destroy all since
+   * some of the params are string literals and we don't want to strdup
+   * everything if we don't have to. RabbitMQ basic_publish PHP 7.x will only
+   * strdup server_address, destination_name, and
+   * messaging_destination_routing_key.
+   */
+  nr_free(message_params.server_address);
```

Comment:
I think it could be cleaner to have a matching macro to `ENSURE_PERSISTENCE` that could be used to un-persist these values in the appropriate way for the PHP is use.

Reply:
Great idea!
[71e5e02](https://github.com/newrelic/newrelic-php-agent/pull/1009/commits/71e5e02e7565d93c37274ae1a45b7e77d0f6779a)